src/database/article.service.ts

```diff
@@ -11,10 +11,67 @@ export class ArticleService {
   ) {}
 
   async findAll(
-    filter?: object,
-    options: QueryOptions = { limit: 30 },
+    filter: { 
+      keyword?: string 
+      tag?: string,
+      category?: string,
+    } = {}, 
+    options: QueryOptions = { limit: 30 }
   ): Promise<Article[]> {
-    return this.articleModel.find(filter ?? {}, null, options).exec();
+    try {
+      let query = {};
+  
+      console.log(
+          'filter:',
+          'keyword:', filter.keyword,
+          'tag:', filter.tag,
+          'category:', filter.category,
+          'options:', options
+      );
+  
+      if (filter.keyword) {
+        query = {
+          $or: [
+            { title: { $regex: filter.keyword, $options: 'i' } },
+          ],
+        };
+      } else if (filter.tag) {
+        query = { tags: filter.tag };
+      } else if (filter.category) {
+        query = { category: filter.category };
+      }
+  
+      const count = await this.articleModel.countDocuments(query);
+  
+      console.log('Number of articles matched with the keyword:', count);
+  
+      // limit articles
+      let finalQuery = this.articleModel.find(query);
+      if (options.limit) {
+        finalQuery = finalQuery.limit(options.limit);
+      }
+  
+      // descending order
+      finalQuery = finalQuery.sort({ _id: -1 });
+  
+      const articles = await finalQuery.exec();
+  
+      // if empty
+      if (!articles || articles.length === 0) {
+        console.log('No articles found with the matching keyword!');
+      }
+  
+      //print article link and title in terminal
+      articles.forEach(article => {
+        console.log('Title:', article.title);
+        console.log('Link:', `https://sync.muilab.org/#/article/${article._id}`);
+      });
+  
+      return articles;
+    } catch (error) {
+      // error if failed to get article
+      console.log('Article Not Found');
+    }
```

Comment:
I believe services should be as simple and general as possible, so it might be best to retain the original implementation. Just allow controllers to define their `filter` and `option`, and handle the result.